cmd/charm-ingest/whitelist.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the GPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/juju/charmstore-client/internal/ingest"
+	"gopkg.in/juju/charmrepo.v4/csclient/params"
+)
+
+var validChannels = params.ValidChannels
+
+// parseWhitelist takes an io.Reader for a whitelist and returns []ingest.WhitelistEntity
```

Comment:
// parseWhitelist parses a whitelist of charms and bundles read from r
	// and returns a slice containing all of them.
	//
	// One entity is defined per line, with the entity id first, followed
	// by all the channels that should be considered for the entity,
	// separated by white space.
	//
	// If the entity id has no specified revision, the most recent published
	// revision for the channels will be specified.

?